### PR TITLE
Adding WDL and Docker Contribution Guides

### PR DESCRIPTION
**Pull request by tefirman**
As the WILDS WDL pipelines prepare to make their debut in a week and a half, we need guidelines around how WDL's and their corresponding Docker images should be designed. This pull request adds initial versions for each set of guidelines and incorporates them into the overarching WILDS Contribution Guide after the "Package Maintenance" section but before the "Security" section. The location of these guides is definitely flexible, just wanted to get an initial version out there and we can move it as needed in the future. These are definitely first drafts, so if you'd like to add/remove guidelines or you feel that something's misworded/misarranged, please feel free to suggest those changes. Thanks for your input! Really appreciate it! #IntoTheWildsWeGo :wolf:

### CODE REVIEW

File: wdlconfig.qmd [hunk 1]
```diff
@@ -0,0 +1,58 @@
+
+# WDL Configuration Guide {{< iconify file-icons wdl >}} {#sec-wdlconfig}
+
+So as not to reinvent the wheel, WILDS WDL's should follow guidelines similar to the ones provided by [BioWDL](https://biowdl.github.io/styleGuidelines.html) and [WARP](https://broadinstitute.github.io/warp/docs/Best_practices/suggested_formats). However, because of the pedagogical "proof-of-concept" nature of the WILDS, they will not be identical and even differ significantly in a few places.
```
Comment (sitapriyamoorthi) [inline wdlconfig.qmd:4]: This is just my personal opinion but starting of with saying "So as not to reinvent the wheel..." Might give the reviewer the impression that okay so this  WDL pipelines simply follows these two guidelines. My take would be just to be upfront and say something like  the WDL configuration guide is inspired by the BioWDL and WARP guidelines.

File: wdlconfig.qmd [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+## WILDS Philosophy
+
+- The mindset behind WILDS is for each repository to be a self-contained demonstration of a particular bioinformatic functionality:
```
Comment (sitapriyamoorthi) [inline wdlconfig.qmd:8]: Is this the philosophy of WILD or WDL WILDS?
Alternative sentence idea:
Each WILDS repository is ideally a self-contained demonstration of a specific bioinformatic workflow.

File: wdlconfig.qmd [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    1. Researcher scans the workflow to deem whether it is relevant to their needs.
+    2. Researcher clones the repository as is, makes minimal updates to the inputs, and easily executes the code locally or otherwise.
+    3. Researcher forks the repository and customizes it as necessary to fit their exact research needs.
```
Comment (sitapriyamoorthi) [inline wdlconfig.qmd:11]: A little confused about these steps. Like are these what we expect the user to do? My take would be that we add a little more information. Something perhaps like this:
1. Each repository will have a Readme file that will detail the over-arching goals of that workflow (for example this workflow will help you convert a fastq file into a cram file). Provide some details about what type of file sources are expected etc. basically everything a user needs to know about the workflow in order to decide if they want to fork and run the workflow. 
2. Once you have decided that this repository is useful to you, you can clone the repository as is, specify your inputs and execute the workflow. 
3. You can also take inspiration from a repository and create a new workflow and then submit it back to the WILDS for review and inclusion.

File: wdlconfig.qmd [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+- To that end, WILDS WDL repositories will usually consist of a single WDL script containing the workflow as well as the tasks that make up the workflow.
+    - This contradicts the recommendations from BioWDL, i.e. tasks should be written in a separate script and imported into the workflow script as a module.
```
Comment (sitapriyamoorthi) [inline wdlconfig.qmd:13]: My take is not to keep referring back to the BioWDL or WARP recommendations. But just to be very straight about what our rules are and this is what you should expect.

File: wdlconfig.qmd [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+So as not to reinvent the wheel, WILDS WDL's should follow guidelines similar to the ones provided by [BioWDL](https://biowdl.github.io/styleGuidelines.html) and [WARP](https://broadinstitute.github.io/warp/docs/Best_practices/suggested_formats). However, because of the pedagogical "proof-of-concept" nature of the WILDS, they will not be identical and even differ significantly in a few places.
+
+## WILDS Philosophy
+
+- The mindset behind WILDS is for each repository to be a self-contained demonstration of a particular bioinformatic functionality:
+    1. Researcher scans the workflow to deem whether it is relevant to their needs.
+    2. Researcher clones the repository as is, makes minimal updates to the inputs, and easily executes the code locally or otherwise.
+    3. Researcher forks the repository and customizes it as necessary to fit their exact research needs.
+- To that end, WILDS WDL repositories will usually consist of a single WDL script containing the workflow as well as the tasks that make up the workflow.
```
Comment (sitapriyamoorthi) [inline wdlconfig.qmd:12]: Maybe add a section on what the user should expect. For example... when you enter a WILDS WDL repo you will find:
1. A readme
2. A WDL
3. An example JSON 
4. Example/test inputs (optional)

**Review by sitapriyamoorthi — CHANGES_REQUESTED** [review]
Hey I think it would just be easier for me to pull a branch make recommendations and then send them over.

File: wdlconfig.qmd (hunk 1 above)
Comment (tefirman) [reply wdlconfig.qmd:4]: Totally agree, it starts things off on the wrong foot. Even if we are very similar to the other two, it's best to establish the guidelines as our own. I'll get that updated right now.

File: wdlconfig.qmd (hunk 2 above)
Comment (tefirman) [reply wdlconfig.qmd:8]: Good point, probably best to be more specific in saying "WILDS WDLs" rather than WILDS in general.

**Comment by tefirman** [comment]
@sitapriyamoorthi -- I pushed some updates to address your comments on the WDL Config file, would you mind taking a look one more time?

File: docker.qmd [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+```
+- When creating a different version of an existing image, use one of the other Dockerfiles as a starting template and modify it as needed.
+    - This will help to ensure that the only thing that has changed between image versions is the version of tool in question, not any strange formatting/configuration issues.
+- Try to be as specific as possible in terms of tool versions within the Dockerfile, especially the parent image.
+    - If you just specify "latest", a tag that get updated frequently over time, your image could be completely different the next time you build it, even though it uses the exact same Dockerfile.
+    - On the other hand, specifying "v1.2.3" will always pull the same instance of the tool every time, providing greater reproducibility over time.
+
+## Repository Guidelines
+
+- In terms of the repo organization, each image should have its own directory named after the tool being used in the image. Each version of the image should have its own Dockerfile in that directory following the naming convention of `[IMAGENAME]/Dockerfile_[VERSIONTAG]`.
```
Comment (laderast) [inline docker.qmd:35]: This is a lot more work, but I think that having examples might be really helpful to some of the more abstract guidelines. At least with relevant lines, and you can point towards code examples in getwilds.

Comment (tefirman) [reply docker.qmd:35]: Great call, I think links to code examples in getWILDS would be perfect, especially for the WDL guidelines.

**Review by sitapriyamoorthi — APPROVED** [review]

**Comment by tefirman** [comment]
Great call, @caalo, thanks for catching that! Just added a few bulletpoints to the Stylistic Guidelines section to address that and added a couple more links out to the WDL101 course in case people want to learn more.

**Pull request merged**